src/dpl/src/Grid.cpp

```diff
@@ -188,6 +188,72 @@ void Opendp::visitCellPixels(
   }
 }
 
+void Opendp::visitCellBoundaryPixels(
+    Cell& cell,
+    bool padded,
+    const std::function<void(Pixel* pixel, Edge edge, int x, int y)>& visitor)
+    const
+{
+  dbInst* inst = cell.db_inst_;
+  dbMaster* master = inst->getMaster();
+  auto obstructions = master->getObstructions();
+  bool have_obstructions = false;
+  for (dbBox* obs : obstructions) {
+    if (obs->getTechLayer()->getType()
+        == odb::dbTechLayerType::Value::OVERLAP) {
+      have_obstructions = true;
+
+      Rect rect = obs->getBox();
+      dbTransform transform;
+      inst->getTransform(transform);
+      transform.apply(rect);
+      int x_start = gridX(rect.xMin() - core_.xMin());
+      int x_end = gridEndX(rect.xMax() - core_.xMin());
+      int y_start = gridY(rect.yMin() - core_.yMin());
+      int y_end = gridEndY(rect.yMax() - core_.yMin());
+      for (int x = x_start; x < x_end; x++) {
+        Pixel* pixel = gridPixel(x, y_start);
+        if (pixel)
+          visitor(pixel, Edge::top, x, y_start);
+        pixel = gridPixel(x, y_end - 1);
+        if (pixel)
+          visitor(pixel, Edge::bottom, x, y_end - 1);
+      }
+      for (int y = y_start; y < y_end; y++) {
+        Pixel* pixel = gridPixel(x_start, y);
+        if (pixel)
+          visitor(pixel, Edge::left, x_start, y);
+        pixel = gridPixel(x_end - 1, y);
+        if (pixel)
+          visitor(pixel, Edge::right, x_end - 1, y);
+      }
```

Comment:
There is no need to visit the top & bottom edges

Reply:
I am a bit confused here. I am looping over the y values, which I imagine would be the left and right edges. Is it the other way around?

I get the part that we don't need to check inner parts of the edge as the corners are enough.

Reply:
Considering that this is a generic function that allows its users to iterate over the boundaries, wouldn't it be useful to leave both? the lambda (visitor) function can decide to ignore top and bottom edges. or we never need to visit top and bottom edges?

Reply:
I don't see a use case for it today but maybe it could be useful in the future.  In that case I would use Direction2D  to indicate the edge